### PR TITLE
#34161: Keep a weak reference to orphan subclasses to reuse on class reload

### LINKED ISSUE #30557: Inner classes behave erratically depending on their parent script's existence in memory
Godot 3.1 Stable

When you load a script into a variable and then load the script again into a new variable, those variables share the same script.

```gdscript
var a = load("res://a.gd")
var b = load("res://a.gd")
# Outputs true due to same shared reference
print(a == b)
```

However if you don't store them anywhere they get unreferenced creating a different identity for the new one.

```gdscript
# These have different identities unlike the first one
print(load("res://a.gd"))
print(load("res://a.gd"))
```

This ends up creating a situation with inner classes that end up behaving erratically depending on whether or not their parent script exists in memory.

If you store the parent scripts in variables and load the inner classes from the stored variables, everything works fine.

```gdscript
var a = load("res://a.gd")
var b = load("res://a.gd")
var a_inner = b.A
var b_inner = b.A
# This will output true because the parent classes still exist in memory
print(a_inner == b_inner)
```

However if you load the inner classes directly, you instance a new parent script every time with a different identity causing the inner classes to also have different identities.

```gdscript
var a_inner = load("res://a.gd").A
var b_inner = load("res://a.gd").A
# This will be false because they have different instances of parent scripts.
print(a_inner == b_inner)
```

I understand why this happens but it doesn't sit right with me from a semantic point of view. As far as I'm concerned as a user I've already loaded A into memory so I should be getting a reference to the same A back. If I loaded a class once and stored it, I loaded a class once and stored it. Otherwise it isn't acting as a class*** but rather just a fancy instance variable. This can also run into issues when we want to use the "instance is class" expression.

***More accurately, a script since that is what inner classes are claimed to be.

### LINKED ISSUE #34161: Assigned type (Class) doesn't match the function argument's type (Class)
Godot 3.2 beta2

I am getting a strange error with typed GDScript, where it finds incompatibility between argument type and parameter type... while they are actually the same.
```
SCRIPT ERROR: GDScript::reload: Parse Error: At "update_episode_data_from_file()" call, argument 1. Assigned type (Project) doesn't match the function argument's type (Project).
          At: res://main.gd:294
```

I have a `script_data.gd` file with a `Project` class inside.

I also have a `script_parser.gd` file which const-imports the previous one, in order to define the following function:
```gdscript
const ScriptData = preload("./script_data.gd")
   ...
static func update_episode_data_from_file(project: ScriptData.Project, path: String) -> Array:
   ...
```

 Then I const-imported both scripts inside `main.gd`, in which I have the following:
```gdscript
const ScriptData = preload("./script_data.gd")
const ScriptParser = preload("./script_parser.gd")
    ...
var _project : ScriptData.Project = null
    ...
func _open_project(fpath: String):
    ...
	var dir := fpath.get_base_dir()
	for ep_file_rpath in data.episode_files:
		var path := dir.plus_file(ep_file_rpath) as String
		ScriptParser.update_episode_data_from_file(_project, path)
```

This code works fine, but recently I've been getting this typing error in the editor, and it obscures any error happening elsewhere after this  line...
So far the only thing that fixes it is to close the editor and reopen it. Then a few hours later, it reappeared again.
I tried to repro in a simpler project, but could not find a pattern yet...

### PR DESCRIPTION
**Pull request by ChibiDenDen**
This change is a second attempt to fix #34161
first attempt is here: https://github.com/godotengine/godot/pull/35027

*Bugsquad edit:* also fixes #30557.

In this change, when a GDScript is destroyed, an ObjectID is kept on GDScriptLanguage for each subclass that is still in use.
when the GDScript is loaded again, it checks if an orphan subclass is still available and reuses it.

To be considered:
the ObjectID is kept in a Map where the Key is a string in the format   CLASS:SUBCLASS, this means that there is no support for multiple levels of subclasses, only one.
IMO, every class should keep a fully qualified name, or some other UNIQUE_ID that is stable and reproducible without requiring a full load of the subclass.

### CODE REVIEW

**Comment by ChibiDenDen** [comment]
I have tested and this seems to fix #30557 as well

**Comment by akien-mga** [comment]
See also #35096 which is @vnen's attempt at solving the same issue (I don't know if it addresses #30557) too.

**Comment by vnen** [comment]
This is better but I fear it has the same problem as the other PR: references are kept alive forever. The difference here is that only some inner classes will be kept. But once they are in that list, they'll only be freed when the application ends.

Maybe we can do a similar thing but only in smaller contexts. For instance, keep this list for the parser while it's running. Or for the compiler it can keep a list of external references in the actual GDScript, so it can take them away when it's freed.

Still, I'm not confident in merging such a solution for 3.2 this close to the release.

**Comment by ChibiDenDen** [comment]
Why do you think that they are kept alive forever?
this is just the objectid, this does not keep reference.
I can create a test case that shows that.

**Comment by vnen** [comment]
I see, I should have another coffee before reading the code.

I guess this is a good solution, since it avoids recreating the inner classes when the owner script is reloaded, so the memory address don't change.

File: modules/gdscript/gdscript.cpp [hunk 1]
```diff
@@ -915,14 +915,43 @@ GDScript::GDScript() :
 #endif
 }
 
+void GDScript::_orphan_subclasses() {
```
Comment (vnen) [inline modules/gdscript/gdscript.cpp]: I think this name can be better, it's not clear what it does. Maybe `_save_orphaned_subclasses()`?

File: modules/gdscript/gdscript.cpp [hunk 2]
```diff
@@ -2176,6 +2205,22 @@ GDScriptLanguage::~GDScriptLanguage() {
 	singleton = NULL;
 }
 
+void GDScriptLanguage::add_orphan_subclass(const String &p_compund_name, const ObjectID &p_subclass) {
```
Comment (vnen) [inline modules/gdscript/gdscript.cpp]: `compund` should be `compound`. But since we are introducing FQN, maybe we want to call it `p_qualified_name` to keep consistency (the same applies to other instances of this argument name).

**Review by vnen — CHANGES_REQUESTED** [review]

**Comment by vnen** [comment]
Also, #35122 should be merged first, so this PR can be updated to use the fully qualified names.

**Comment by vnen** [comment]
Looks good, thanks. Just need to squash the commits into one. See http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase

**Comment by ChibiDenDen** [comment]
> Looks good, thanks. Just need to squash the commits into one. See http://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#mastering-the-pr-workflow-the-rebase

Done, its a single commit now

File: modules/gdscript/gdscript_compiler.cpp [hunk 3]
```diff
@@ -2123,15 +2123,21 @@ void GDScriptCompiler::_make_scripts(GDScript *p_script, const GDScriptParser::C
 		StringName name = p_class->subclasses[i]->name;
 
 		Ref<GDScript> subclass;
+		String fully_qualified_name = p_script->fully_qualified_name + "::" + name;
```
Comment (akien-mga) [inline modules/gdscript/gdscript_compiler.cpp]: You might need to rebase on the master branch, as those changes should not be part of this PR since they were already merged.

Comment (ChibiDenDen) [reply modules/gdscript/gdscript_compiler.cpp]: But this line isnt part of the PR, in master subclass->fully_qualified_name is initiated directly, in this commit i moved it to a local variable to reuse the value instead of calculating it twice.

**Review by vnen — APPROVED** [review]
Looks good. It's a clever solution while we don't have the proper ScriptCache.

**Comment by akien-mga** [comment]
Thanks!
